File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/internal/EventSystemTest.java
package dev.robocode.tankroyale.botapi.internal;

import dev.robocode.tankroyale.botapi.*;
import dev.robocode.tankroyale.botapi.events.*;
import dev.robocode.tankroyale.botapi.graphics.Color;
import dev.robocode.tankroyale.botapi.graphics.IGraphics;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.util.*;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;

@Tag("EVT")
class EventSystemTest {

    static class BaseBotStub implements IBaseBot {
        final List<BotEvent> firedEvents = new ArrayList<>();

        @Override public void start() {}
        @Override public void go() {}
        @Override public int getMyId() { return 0; }
        @Override public String getVariant() { return ""; }
        @Override public String getVersion() { return ""; }
        @Override public String getGameType() { return ""; }
        @Override public int getArenaWidth() { return 0; }
        @Override public int getArenaHeight() { return 0; }
        @Override public int getNumberOfRounds() { return 0; }
        @Override public double getGunCoolingRate() { return 0; }
        @Override public int getMaxInactivityTurns() { return 0; }
        @Override public int getTurnTimeout() { return 0; }
        @Override public int getTimeLeft() { return 0; }
        @Override public int getRoundNumber() { return 0; }
        @Override public int getTurnNumber() { return 0; }
        @Override public int getEnemyCount() { return 0; }
        @Override public double getEnergy() { return 0; }
        @Override public boolean isDisabled() { return false; }
        @Override public double getX() { return 0; }
        @Override public double getY() { return 0; }
        @Override public double getDirection() { return 0; }
        @Override public double getGunDirection() { return 0; }
        @Override public double getRadarDirection() { return 0; }
        @Override public double getSpeed() { return 0; }
        @Override public double getGunHeat() { return 0; }
        @Override public Collection<BulletState> getBulletStates() { return Collections.emptyList(); }
        @Override public List<BotEvent> getEvents() { return Collections.emptyList(); }
        @Override public void clearEvents() {}
        @Override public double getTurnRate() { return 0; }
        @Override public void setTurnRate(double turnRate) {}
        @Override public double getMaxTurnRate() { return 0; }
        @Override public void setMaxTurnRate(double maxTurnRate) {}
        @Override public double getGunTurnRate() { return 0; }
        @Override public void setGunTurnRate(double gunTurnRate) {}
        @Override public double getMaxGunTurnRate() { return 0; }
        @Override public void setMaxGunTurnRate(double maxGunTurnRate) {}
        @Override public double getRadarTurnRate() { return 0; }
        @Override public void setRadarTurnRate(double radarTurnRate) {}
        @Override public double getMaxRadarTurnRate() { return 0; }
        @Override public void setMaxRadarTurnRate(double maxRadarTurnRate) {}
        @Override public double getTargetSpeed() { return 0; }
        @Override public void setTargetSpeed(double targetSpeed) {}
        @Override public double getMaxSpeed() { return 0; }
        @Override public void setMaxSpeed(double maxSpeed) {}
        @Override public boolean setFire(double firepower) { return false; }
        @Override public double getFirepower() { return 0; }
        @Override public void setRescan() {}
        @Override public void setFireAssist(boolean enable) {}
        @Override public void setInterruptible(boolean interruptible) {}
        @Override public void setAdjustGunForBodyTurn(boolean adjust) {}
        @Override public boolean isAdjustGunForBodyTurn() { return false; }
        @Override public void setAdjustRadarForBodyTurn(boolean adjust) {}
        @Override public boolean isAdjustRadarForBodyTurn() { return false; }
        @Override public void setAdjustRadarForGunTurn(boolean adjust) {}
        @Override public boolean isAdjustRadarForGunTurn() { return false; }
        @Override public boolean addCustomEvent(Condition condition) { return false; }
        @Override public boolean removeCustomEvent(Condition condition) { return false; }
        @Override public void setStop() {}
        @Override public void setStop(boolean overwrite) {}
        @Override public void setResume() {}
        @Override public Set<Integer> getTeammateIds() { return Collections.emptySet(); }
        @Override public boolean isTeammate(int botId) { return false; }
        @Override public void broadcastTeamMessage(Object message) {}
        @Override public void sendTeamMessage(int teammateId, Object message) {}
        @Override public boolean isStopped() { return false; }
        @Override public Color getBodyColor() { return null; }
        @Override public void setBodyColor(Color color) {}
        @Override public Color getTurretColor() { return null; }
        @Override public void setTurretColor(Color color) {}
        @Override public Color getRadarColor() { return null; }
        @Override public void setRadarColor(Color color) {}
        @Override public Color getBulletColor() { return null; }
        @Override public void setBulletColor(Color color) {}
        @Override public Color getScanColor() { return null; }
        @Override public void setScanColor(Color color) {}
        @Override public Color getTracksColor() { return null; }
        @Override public void setTracksColor(Color color) {}
        @Override public Color getGunColor() { return null; }
        @Override public void setGunColor(Color color) {}
        @Override public boolean isDebuggingEnabled() { return false; }
        @Override public IGraphics getGraphics() { return null; }
        @Override public double calcMaxTurnRate(double speed) { return 0; }
        @Override public double calcBulletSpeed(double firepower) { return 0; }
        @Override public double calcGunHeat(double firepower) { return 0; }
        @Override public int getEventPriority(Class<BotEvent> eventClass) { return 0; }
        @Override public void setEventPriority(Class<BotEvent> eventClass, int priority) {}

        @Override public void onConnected(ConnectedEvent e) {}
        @Override public void onDisconnected(DisconnectedEvent e) {}
        @Override public void onConnectionError(ConnectionErrorEvent e) {}
        @Override public void onGameStarted(GameStartedEvent e) {}
        @Override public void onGameEnded(GameEndedEvent e) {}
        @Override public void onRoundStarted(RoundStartedEvent e) {}
        @Override public void onRoundEnded(RoundEndedEvent e) {}
        @Override public void onTick(TickEvent e) { firedEvents.add(e); }
        @Override public void onBotDeath(BotDeathEvent e) { firedEvents.add(e); }
        @Override public void onDeath(DeathEvent e) { firedEvents.add(e); }
        @Override public void onHitBot(HitBotEvent e) { firedEvents.add(e); }
        @Override public void onHitWall(HitWallEvent e) { firedEvents.add(e); }
        @Override public void onBulletFired(BulletFiredEvent e) { firedEvents.add(e); }
        @Override public void onHitByBullet(HitByBulletEvent e) { firedEvents.add(e); }
        @Override public void onBulletHit(BulletHitBotEvent e) { firedEvents.add(e); }
        @Override public void onBulletHitBullet(BulletHitBulletEvent e) { firedEvents.add(e); }
        @Override public void onBulletHitWall(BulletHitWallEvent e) { firedEvents.add(e); }
        @Override public void onScannedBot(ScannedBotEvent e) { firedEvents.add(e); }
        @Override public void onWonRound(WonRoundEvent e) { firedEvents.add(e); }
        @Override public void onCustomEvent(CustomEvent e) { firedEvents.add(e); }
        @Override public void onTeamMessage(TeamMessageEvent e) { firedEvents.add(e); }
    }

    private BaseBotStub botStub;
    private BaseBotInternals internals;
    private EventQueue queue;

    @BeforeEach
    void setUp() {
        botStub = new BaseBotStub();
        BotInfo botInfo = new BotInfo("dummy", "1.0", List.of("dummy"), null, null, null, null, null, null, null);
        internals = new BaseBotInternals(botStub, botInfo, URI.create("ws://localhost:7654"), null);
        queue = new EventQueue(internals, internals.getBotEventHandlers());
    }

    @Test
    @Tag("TR-API-EVT-001")
    void test_TR_API_EVT_001_event_constructors() {
        // TickEvent
        TickEvent te = new TickEvent(1, 2, null, Collections.emptyList(), Collections.emptyList());
        assertEquals(1, te.getTurnNumber());
        assertEquals(2, te.getRoundNumber());

        // ScannedBotEvent
        ScannedBotEvent sbe = new ScannedBotEvent(3, 1, 2, 80.0, 100.0, 200.0, 45.0, 5.0);
        assertEquals(3, sbe.getTurnNumber());
        assertEquals(1, sbe.getScannedByBotId());
        assertEquals(2, sbe.getScannedBotId());
        assertEquals(80.0, sbe.getEnergy());
        assertEquals(100.0, sbe.getX());
        assertEquals(200.0, sbe.getY());
        assertEquals(45.0, sbe.getDirection());
        assertEquals(5.0, sbe.getSpeed());

        // HitBotEvent
        HitBotEvent hbe = new HitBotEvent(4, 5, 90.0, 10.0, 20.0, true);
        assertEquals(4, hbe.getTurnNumber());
        assertEquals(5, hbe.getVictimId());
        assertEquals(90.0, hbe.getEnergy());
        assertEquals(10.0, hbe.getX());
        assertEquals(20.0, hbe.getY());
        assertTrue(hbe.isRammed());

        // HitByBulletEvent
        BulletState bullet = new BulletState(1, 1, 3.0, 100.0, 200.0, 45.0, null);
        HitByBulletEvent hbbe = new HitByBulletEvent(6, bullet, 5.0, 95.0);
        assertEquals(6, hbbe.getTurnNumber());
        assertEquals(bullet, hbbe.getBullet());
        assertEquals(5.0, hbbe.getDamage());
        assertEquals(95.0, hbbe.getEnergy());

        // HitWallEvent
        HitWallEvent hwe = new HitWallEvent(7);
        assertEquals(7, hwe.getTurnNumber());

        // BulletFiredEvent
        BulletFiredEvent bfe = new BulletFiredEvent(8, bullet);
        assertEquals(8, bfe.getTurnNumber());
        assertEquals(bullet, bfe.getBullet());

        // BulletHitBotEvent
        BulletHitBotEvent bhbe = new BulletHitBotEvent(9, 10, bullet, 5.0, 90.0);
        assertEquals(9, bhbe.getTurnNumber());
        assertEquals(bullet, bhbe.getBullet());
        assertEquals(10, bhbe.getVictimId());
        assertEquals(5.0, bhbe.getDamage());
        assertEquals(90.0, bhbe.getEnergy());

        // BulletHitBulletEvent
        BulletState otherBullet = new BulletState(2, 2, 3.0, 150.0, 250.0, 90.0, null);
        BulletHitBulletEvent bhbue = new BulletHitBulletEvent(10, bullet, otherBullet);
        assertEquals(10, bhbue.getTurnNumber());
        assertEquals(bullet, bhbue.getBullet());
        assertEquals(otherBullet, bhbue.getHitBullet());

        // BulletHitWallEvent
        BulletHitWallEvent bhwe = new BulletHitWallEvent(11, bullet);
        assertEquals(11, bhwe.getTurnNumber());
        assertEquals(bullet, bhwe.getBullet());

        // BotDeathEvent
        BotDeathEvent bde = new BotDeathEvent(12, 13);
        assertEquals(12, bde.getTurnNumber());
        assertEquals(13, bde.getVictimId());

        // DeathEvent
        DeathEvent de = new DeathEvent(14);
        assertEquals(14, de.getTurnNumber());

        // SkippedTurnEvent
        SkippedTurnEvent ste = new SkippedTurnEvent(15);
        assertEquals(15, ste.getTurnNumber());

        // WonRoundEvent
        WonRoundEvent wre = new WonRoundEvent(16);
        assertEquals(16, wre.getTurnNumber());

        // TeamMessageEvent
        TeamMessageEvent tme = new TeamMessageEvent(17, "hello", 18);
        assertEquals(17, tme.getTurnNumber());
        assertEquals("hello", tme.getMessage());
        assertEquals(18, tme.getSenderId());

        // CustomEvent
        Condition condition = new Condition("test") {
            @Override public boolean test() { return true; }
        };
        CustomEvent ce = new CustomEvent(19, condition);
        assertEquals(19, ce.getTurnNumber());
        assertEquals(condition, ce.getCondition());
    }

    @Test
    @Tag("TR-API-EVT-008")
    void test_TR_API_EVT_008_condition_test_callable() {
        Condition c1 = new Condition("true") {
            @Override public boolean test() { return true; }
        };
        assertTrue(c1.test());

        Condition c2 = new Condition("false") {
            @Override public boolean test() { return false; }
        };
        assertFalse(c2.test());
    }

    @Test
    @Tag("TR-API-EVT-009")
    void test_TR_API_EVT_009_custom_event_firing() {
        Condition condTrue = new Condition("true") {
            @Override public boolean test() { return true; }
        };
        Condition condFalse = new Condition("false") {
            @Override public boolean test() { return false; }
        };

        internals.addCondition(condTrue);
        internals.addCondition(condFalse);
        
        TickEvent tick = new TickEvent(5, 1, null, Collections.emptyList(), Collections.emptyList());
        internals.setTickEvent(tick);

        queue.dispatchEvents(5);

        assertThat(botStub.firedEvents).hasSize(1);
        assertThat(botStub.firedEvents.get(0)).isInstanceOf(CustomEvent.class);
        assertEquals(condTrue, ((CustomEvent) botStub.firedEvents.get(0)).getCondition());
    }
}

File: bot-api/dotnet/test/src/internal/EventSystemTest.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using NUnit.Framework;
using Robocode.TankRoyale.BotApi.Events;
using Robocode.TankRoyale.BotApi.Internal;
using Robocode.TankRoyale.BotApi.Graphics;

namespace Robocode.TankRoyale.BotApi.Tests.Internal;

[TestFixture]
[Category("EVT")]
public class EventSystemTest
{
    private class BaseBotStub : IBaseBot
    {
        public List<BotEvent> FiredEvents { get; } = new();

        public void Start() {}
        public void Go() {}
        public int MyId => 1;
        public string Variant => "";
        public string Version => "";
        public string GameType => "";
        public int ArenaWidth => 800;
        public int ArenaHeight => 600;
        public int NumberOfRounds => 1;
        public double GunCoolingRate => 0.1;
        public int? MaxInactivityTurns => 450;
        public int TurnTimeout => 30000;
        public int TimeLeft => 0;
        public int RoundNumber => 1;
        public int TurnNumber => 1;
        public int EnemyCount => 0;
        public double Energy { get; set; } = 100;
        public bool IsDisabled => Energy <= 0;
        public double X => 0;
        public double Y => 0;
        public double Direction => 0;
        public double GunDirection => 0;
        public double RadarDirection => 0;
        public double Speed => 0;
        public double GunHeat { get; set; } = 0;
        public IEnumerable<BulletState> BulletStates => new List<BulletState>();
        public IList<BotEvent> Events => new List<BotEvent>();
        public void ClearEvents() {}
        public double TurnRate { get; set; }
        public double MaxTurnRate { get; set; } = 10;
        public double GunTurnRate { get; set; }
        public double MaxGunTurnRate { get; set; } = 20;
        public double RadarTurnRate { get; set; }
        public double MaxRadarTurnRate { get; set; } = 45;
        public double TargetSpeed { get; set; }
        public double MaxSpeed { get; set; } = 8;
        public bool SetFire(double firepower) => false;
        public double Firepower => 0;
        public void SetRescan() {}
        public void SetFireAssist(bool enable) {}
        public bool Interruptible { set {} }
        public bool AdjustGunForBodyTurn { get; set; }
        public bool AdjustRadarForBodyTurn { get; set; }
        public bool AdjustRadarForGunTurn { get; set; }
        public void SetStop() {}
        public void SetStop(bool overwrite) {}
        public void SetResume() {}
        public bool IsStopped => false;
        public ICollection<int> TeammateIds => new List<int>();
        public bool IsTeammate(int botId) => false;
        public void BroadcastTeamMessage(object message) {}
        public void SendTeamMessage(int teammateId, object message) {}
        public Color? BodyColor { get; set; }
        public Color? TurretColor { get; set; }
        public Color? RadarColor { get; set; }
        public Color? BulletColor { get; set; }
        public Color? ScanColor { get; set; }
        public Color? TracksColor { get; set; }
        public Color? GunColor { get; set; }
        public bool IsDebuggingEnabled => false;
        public IGraphics Graphics => null;

        public void OnConnected(ConnectedEvent e) {}
        public void OnDisconnected(DisconnectedEvent e) {}
        public void OnConnectionError(ConnectionErrorEvent e) {}
        public void OnGameStarted(GameStartedEvent e) {}
        public void OnGameEnded(GameEndedEvent e) {}
        public void OnRoundStarted(RoundStartedEvent e) {}
        public void OnRoundEnded(RoundEndedEvent e) {}
        public void OnTick(TickEvent e) => FiredEvents.Add(e);
        public void OnBotDeath(BotDeathEvent e) => FiredEvents.Add(e);
        public void OnDeath(DeathEvent e) => FiredEvents.Add(e);
        public void OnHitBot(HitBotEvent e) => FiredEvents.Add(e);
        public void OnHitWall(HitWallEvent e) => FiredEvents.Add(e);
        public void OnBulletFired(BulletFiredEvent e) => FiredEvents.Add(e);
        public void OnHitByBullet(HitByBulletEvent e) => FiredEvents.Add(e);
        public void OnBulletHit(BulletHitBotEvent e) => FiredEvents.Add(e);
        public void OnBulletHitBullet(BulletHitBulletEvent e) => FiredEvents.Add(e);
        public void OnBulletHitWall(BulletHitWallEvent e) => FiredEvents.Add(e);
        public void OnScannedBot(ScannedBotEvent e) => FiredEvents.Add(e);
        public void OnSkippedTurn(SkippedTurnEvent e) => FiredEvents.Add(e);
        public void OnWonRound(WonRoundEvent e) => FiredEvents.Add(e);
        public void OnCustomEvent(CustomEvent e) => FiredEvents.Add(e);
        public void OnTeamMessage(TeamMessageEvent e) => FiredEvents.Add(e);

        public bool AddCustomEvent(Condition condition) => false;
        public bool RemoveCustomEvent(Condition condition) => false;
        public int GetEventPriority(Type t) => 0;
        public void SetEventPriority(Type t, int p) {}
        public double CalcMaxTurnRate(double s) => 0;
        public double CalcBulletSpeed(double f) => 0;
        public double CalcGunHeat(double f) => 0;
        public double CalcBearing(double d) => 0;
        public double CalcGunBearing(double d) => 0;
        public double CalcRadarBearing(double d) => 0;
        public double DirectionTo(double x, double y) => 0;
        public double BearingTo(double x, double y) => 0;
        public double GunBearingTo(double x, double y) => 0;
        public double RadarBearingTo(double x, double y) => 0;
        public double DistanceTo(double x, double y) => 0;
        public double NormalizeAbsoluteAngle(double a) => 0;
        public double NormalizeRelativeAngle(double a) => 0;
        public double CalcDeltaAngle(double t, double s) => 0;
    }

    private BaseBotStub botStub;
    private BaseBotInternals internals;
    private EventQueue queue;

    [SetUp]
    public void SetUp()
    {
        botStub = new BaseBotStub();
        var botInfo = new BotInfo("dummy", "1.0", new List<string> { "dummy" }, null, null, null, null, null, null, null);
        internals = new BaseBotInternals(botStub, botInfo, new Uri("ws://localhost:7654"), null);
        queue = new EventQueue(internals, internals.BotEventHandlers);
    }

    [Test]
    [Property("ID", "TR-API-EVT-001")]
    public void Test_TR_API_EVT_001_Event_Constructors()
    {
        // TickEvent
        var te = new TickEvent(1, 2, null, new List<BulletState>(), new List<BotEvent>());
        Assert.That(te.TurnNumber, Is.EqualTo(1));
        Assert.That(te.RoundNumber, Is.EqualTo(2));

        // ScannedBotEvent
        var sbe = new ScannedBotEvent(3, 1, 2, 80.0, 100.0, 200.0, 45.0, 5.0);
        Assert.That(sbe.TurnNumber, Is.EqualTo(3));
        Assert.That(sbe.ScannedByBotId, Is.EqualTo(1));
        Assert.That(sbe.ScannedBotId, Is.EqualTo(2));
        Assert.That(sbe.Energy, Is.EqualTo(80.0));
        Assert.That(sbe.X, Is.EqualTo(100.0));
        Assert.That(sbe.Y, Is.EqualTo(200.0));
        Assert.That(sbe.Direction, Is.EqualTo(45.0));
        Assert.That(sbe.Speed, Is.EqualTo(5.0));

        // HitBotEvent
        var hbe = new HitBotEvent(4, 5, 90.0, 10.0, 20.0, true);
        Assert.That(hbe.TurnNumber, Is.EqualTo(4));
        Assert.That(hbe.VictimId, Is.EqualTo(5));
        Assert.That(hbe.Energy, Is.EqualTo(90.0));
        Assert.That(hbe.X, Is.EqualTo(10.0));
        Assert.That(hbe.Y, Is.EqualTo(20.0));
        Assert.That(hbe.IsRammed, Is.True);

        // HitByBulletEvent
        var bullet = new BulletState(1, 1, 3.0, 100.0, 200.0, 45.0, null);
        var hbbe = new HitByBulletEvent(6, bullet, 5.0, 95.0);
        Assert.That(hbbe.TurnNumber, Is.EqualTo(6));
        Assert.That(hbbe.Bullet, Is.EqualTo(bullet));
        Assert.That(hbbe.Damage, Is.EqualTo(5.0));
        Assert.That(hbbe.Energy, Is.EqualTo(95.0));

        // HitWallEvent
        var hwe = new HitWallEvent(7);
        Assert.That(hwe.TurnNumber, Is.EqualTo(7));

        // BulletFiredEvent
        var bfe = new BulletFiredEvent(8, bullet);
        Assert.That(bfe.TurnNumber, Is.EqualTo(8));
        Assert.That(bfe.Bullet, Is.EqualTo(bullet));

        // BulletHitBotEvent
        var bhbe = new BulletHitBotEvent(9, 10, bullet, 5.0, 90.0);
        Assert.That(bhbe.TurnNumber, Is.EqualTo(9));
        Assert.That(bhbe.VictimId, Is.EqualTo(10));
        Assert.That(bhbe.Bullet, Is.EqualTo(bullet));
        Assert.That(bhbe.Damage, Is.EqualTo(5.0));
        Assert.That(bhbe.Energy, Is.EqualTo(90.0));

        // BulletHitBulletEvent
        var otherBullet = new BulletState(2, 2, 3.0, 150.0, 250.0, 90.0, null);
        var bhbue = new BulletHitBulletEvent(10, bullet, otherBullet);
        Assert.That(bhbue.TurnNumber, Is.EqualTo(10));
        Assert.That(bhbue.Bullet, Is.EqualTo(bullet));
        Assert.That(bhbue.HitBullet, Is.EqualTo(otherBullet));

        // BulletHitWallEvent
        var bhwe = new BulletHitWallEvent(11, bullet);
        Assert.That(bhwe.TurnNumber, Is.EqualTo(11));
        Assert.That(bhwe.Bullet, Is.EqualTo(bullet));

        // BotDeathEvent
        var bde = new BotDeathEvent(12, 13);
        Assert.That(bde.TurnNumber, Is.EqualTo(12));
        Assert.That(bde.VictimId, Is.EqualTo(13));

        // DeathEvent
        var de = new DeathEvent(14);
        Assert.That(de.TurnNumber, Is.EqualTo(14));

        // SkippedTurnEvent
        var ste = new SkippedTurnEvent(15);
        Assert.That(ste.TurnNumber, Is.EqualTo(15));

        // WonRoundEvent
        var wre = new WonRoundEvent(16);
        Assert.That(wre.TurnNumber, Is.EqualTo(16));

        // TeamMessageEvent
        var tme = new TeamMessageEvent(17, "hello", 18);
        Assert.That(tme.TurnNumber, Is.EqualTo(17));
        Assert.That(tme.Message, Is.EqualTo("hello"));
        Assert.That(tme.SenderId, Is.EqualTo(18));

        // CustomEvent
        var condition = new TestCondition();
        var ce = new CustomEvent(19, condition);
        Assert.That(ce.TurnNumber, Is.EqualTo(19));
        Assert.That(ce.Condition, Is.EqualTo(condition));
    }

    [Test]
    [Property("ID", "TR-API-EVT-005")]
    public void Test_TR_API_EVT_005_Event_Queue_Priority()
    {
        var wre = new WonRoundEvent(1); // Critical, Priority 150
        var de = new DeathEvent(1);     // Critical, Priority 10
        var sbe = new ScannedBotEvent(1, 1, 2, 100, 100, 100, 0, 0); // Non-critical, Priority 20

        queue.AddEvent(de);
        queue.AddEvent(wre);
        queue.AddEvent(sbe);

        queue.DispatchEvents(1);

        Assert.That(botStub.FiredEvents, Has.Count.EqualTo(3));
        Assert.That(botStub.FiredEvents[0], Is.InstanceOf<WonRoundEvent>());
        Assert.That(botStub.FiredEvents[1], Is.InstanceOf<DeathEvent>());
        Assert.That(botStub.FiredEvents[2], Is.InstanceOf<ScannedBotEvent>());
    }

    [Test]
    [Property("ID", "TR-API-EVT-006")]
    public void Test_TR_API_EVT_006_Event_Queue_Age_Culling()
    {
        var oldEvent = new ScannedBotEvent(7, 1, 2, 100, 100, 100, 0, 0);
        var fineEvent = new ScannedBotEvent(8, 1, 2, 100, 100, 100, 0, 0);
        var oldCritical = new WonRoundEvent(7);

        queue.AddEvent(oldEvent);
        queue.AddEvent(fineEvent);
        queue.AddEvent(oldCritical);

        queue.DispatchEvents(10);

        Assert.That(botStub.FiredEvents, Has.Count.EqualTo(2));
        Assert.That(botStub.FiredEvents[0], Is.InstanceOf<WonRoundEvent>());
        Assert.That(botStub.FiredEvents[1], Is.InstanceOf<ScannedBotEvent>());
    }

    [Test]
    [Property("ID", "TR-API-EVT-007")]
    public void Test_TR_API_EVT_007_Event_Queue_Size_Cap()
    {
        for (int i = 0; i < 266; i++)
        {
            queue.AddEvent(new SkippedTurnEvent(i));
        }
        Assert.That(queue.Events(300), Has.Count.EqualTo(256));
    }

    [Test]
    [Property("ID", "TR-API-EVT-008")]
    public void Test_TR_API_EVT_008_Condition_Test_Callable()
    {
        var c1 = new TestCondition(true);
        Assert.That(c1.Test(), Is.True);

        var c2 = new TestCondition(false);
        Assert.That(c2.Test(), Is.False);
    }

    [Test]
    [Property("ID", "TR-API-EVT-009")]
    public void Test_TR_API_EVT_009_Custom_Event_Firing()
    {
        var condTrue = new TestCondition(true);
        var condFalse = new TestCondition(false);

        internals.AddCondition(condTrue);
        internals.AddCondition(condFalse);
        
        var tick = new TickEvent(5, 1, null, new List<BulletState>(), new List<BotEvent>());
        SetTickEvent(tick);

        queue.DispatchEvents(5);

        Assert.That(botStub.FiredEvents, Has.Count.EqualTo(1));
        Assert.That(botStub.FiredEvents[0], Is.InstanceOf<CustomEvent>());
        Assert.That(((CustomEvent)botStub.FiredEvents[0]).Condition, Is.EqualTo(condTrue));
    }

    private void SetTickEvent(TickEvent tick)
    {
        var field = typeof(BaseBotInternals).GetField("_tickEvent", BindingFlags.NonPublic | BindingFlags.Instance);
        field.SetValue(internals, tick);
    }

    private class TestCondition : Condition
    {
        private readonly bool _result;
        public TestCondition(bool result = true) : base("test") { _result = result; }
        public override bool Test() => _result;
    }
}
